walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -1,23 +1,189 @@
+                        
+                        
+                    }
+                
+            }
+
+            if(k == busySchedule.size()-1 && busySchedule.size()!=0){
+                
+                TimeRange possibleTime = TimeRange.fromStartEnd(end,dayEnd,true);
+
+                if(k!=0){
+                    int possibleStartTime = Math.max(end,possibleTime.end());
+                    TimeRange anotherPossibleTime = TimeRange.fromStartEnd(end,dayEnd,true);
+
+                    if(anotherPossibleTime.overlaps(possibleTime)){
+                        if(anotherPossibleTime.duration() >= meetingDuration){
+                            possibleTimes.add(anotherPossibleTime);
+                        
+                        }
+
+                    }
+                    
+                }else{
+                    if(possibleTime.duration() >= meetingDuration){
+                        possibleTimes.add(possibleTime); 
+                        
+                    }
+                }     
+            }
+
+            if(k+1<busySchedule.size()){
+                
+                TimeRange possibleTime = TimeRange.fromStartEnd(end,busySchedule.get(k+1).start(),false);
+                
+                    if(possibleTime.duration() >= meetingDuration){
+                    possibleTimes.add(possibleTime);
+
+                    
+                    
+                }
+            }
+
+        }
+        
+
+
+    
+     return possibleTimes;
   }
+
+    //check if person is attendee of event
+   public boolean isAnAttendee(String attendee, Event event){
+       if (event.getAttendees().contains(attendee)){
+            return true;
+       }        
+       else{
+            return false;
+       }
+                
+   }
+    //adds event to an attendee schedule. 
+   public void addToSchedule(HashSet<TimeRange> attendeeSchedule, Event event){
+       attendeeSchedule.add(event.getWhen());
+   }
+    //check if there is an overlap between a meeting time and an attendee schedule
+   public boolean checkIfOverlap(TimeRange currentMeetingTime,HashSet<TimeRange> attendeeSchedule){
+
+     Iterator<TimeRange> i = attendeeSchedule.iterator(); 
+        while (i.hasNext()){
+            if(i.next().overlaps(currentMeetingTime)){
+                return true;
+            }
+            
+        }
+
+        return false;
+            
+   }
+
+   public void trimContainedNodes(List<TimeRange>listThatMayContainNestedEvents){
+       for(int i =0 ; i < listThatMayContainNestedEvents.size() ; i++){//iterating through the list and removing at the same time. (mark)
+           for(int j =0 ; j< listThatMayContainNestedEvents.size(); j++){
+                if(i==j){//skip for same one
+                    continue;
+                }else if(listThatMayContainNestedEvents.get(i).contains(listThatMayContainNestedEvents.get(j))){
+                   listThatMayContainNestedEvents.remove(j);
```

Comment:
If you remove an element from the list while you're iterating over it, won't you potentially skip over the next element?

(Funny enough this is a good place to use an `Iterator` instead of a regular `for` loop!)